src/main/java/io/nats/client/impl/MessageQueue.java

```diff
@@ -27,39 +28,30 @@
     private final static int RUNNING = 1;
     private final static int DRAINING = 2;
 
-    public static final int MAX_SPINS = 200; // Default max spins
-    public static final int SPIN_WAIT = 50;
-    public static final int MAX_SPIN_TIME = SPIN_WAIT * MAX_SPINS;
-
     private final AtomicLong length;
     private final AtomicLong sizeInBytes;
     private final AtomicInteger running;
     private final boolean singleThreadedReader;
-    private final int maxSpins;
-    private final ConcurrentLinkedQueue<NatsMessage> queue;
-    private final ConcurrentLinkedQueue<Thread> waiters;
+    private final LinkedBlockingQueue<NatsMessage> queue;
     private final Lock filterLock;
+    private final NatsMessage poisonPill;
 
-    MessageQueue(boolean singleReaderMode) {
-        this.queue = new ConcurrentLinkedQueue<>();
+    MessageQueue(boolean singleReaderMode, int publishHighwaterMark) {
+        this.queue = publishHighwaterMark > 0 ? new LinkedBlockingQueue<NatsMessage>(publishHighwaterMark) : new LinkedBlockingQueue<NatsMessage>();
         this.running = new AtomicInteger(RUNNING);
         this.sizeInBytes = new AtomicLong(0);
         this.length = new AtomicLong(0);
 
+        // The poisonPill is used to stop poll and accumulate when the queue is stopped
+        this.poisonPill = new NatsMessage("_poison", null, NatsConnection.EMPTY_BODY, false);

```

Comment:
This makes me feel like we want something to crash, do we want that or is this a `pauseSignal` or `blockSignal` or something like that?

Reply:
poisonpill was the descriptive term used by the people that said how to stop a linked blocking queue, so I kept it.

Reply:
Gotcha..